Food irradiator

nexusmrsep opened the pull request:
#### Summary
```SUMMARY: Features "Adds food irradiator computer console and location"```

#### Purpose of change
This feature dates back to #24145
Generally we have irradiated food (food preserved by applying radiation until bacteria die), but we don't have any actual way to irradiate food.

#### Describe the solution
This adds infrastructure to code-in a computer console in new map locations with option to irradiate items, especially food. It's based around real life radiation processing plants/factories. Conveyor belts move food to the radiation platform where radiation source irradiated the food, and then it is moved back into unloading bay. I tried to simulate this process (with some twists for !!fun!!) by adding subsequent functions to the computer console.

[EDIT] Also per request I've built a fully functional facility overmap location with many fun features.

**Features:**
- converts food to it's irradiated version (if such exists)
- spews radiation in the process
- added sub-feature: Geiger counter option to measure radiation on radiation platform and at player's tile
- added sub-feature: conveyor belt's loading and unloading bay item cycling via console options, to operate loading/unloading items on the platform without getting close to the radiation source
- added sub-feature: toggle reinforced glass shutters - beside visual effect of allowing to inspect the radiation chamber, closed shutters block line of sight during radiation cycle, and protect user from direct exposure to radiation
- added some !!fun!! for those who would want to irradiate some certain ~~radioactive~~ items available in game
- added sub-feature: ability to remove radiation source from irradiator
- [EDIT] added sub-feature: power management computer (shock vent disabler)
- [EDIT] added sub-feature: industrial ID card & industrial card reader

SIDE-QUESTS [extras I did that enhance the overall effect]:
- updated translate_radius() function/method to have toggle_between option that switches between two types of terrain (before it only went 1->2, now it optionally does both: 1->2 & 2->1 simultaneously) useful in shutters functions where it toggle between both states (open/closed) via a single computer option
- fixed mininuke timer to say exactly that your input is in turns (no more guessing) - also the confirmation comes textified for ex. "You set the timer to 5 minutes" Much more convenient feedback.
- in Geiger counters I named the number provided using real world unit miliSieverts/hour = mSv/h for ground radiation and mSv for body dose taken. For example radiation badge turns black at 500 mSv. I was wondering if this unit corresponds well to what happens in the game, but according to my little research, it would suffice.

Reference via [Wikipedia](https://en.wikipedia.org/wiki/Sievert):
- 500 | mSv: | the U.S. 10 C.F.R. § 20.1201(a)(2)(ii) occupational dose limit, shallow-dose equivalent to skin, per annum
- 670 | mSv: | highest dose received by a worker responding to the Fukushima emergency
- 1 | Sv: | maximum allowed radiation exposure for NASA astronauts over their career
- 4–5 | Sv: | dose required to kill a human with a 50% risk within 30 days (LD50/30), if the dose is received over a very short duration

All above can be used and manipulated in various ways withing the limitations and constrains depending how you build a map around it. For example leave less space between user and platform and the radiation will break through, etc.

#### Describe alternatives you've considered
- Vehicle part irradiator
- actual irradiated food shelf life (killing bacteria slows decay, but does't stop it completely, chemical decomposition and oxidation is not affected)

#### Additional context
This is my crown jewel of my CDDA development participation, this sheer idea brought me initially to learning coding CDDA until I was able to do such a feat. Enjoy.

![image](https://user-images.githubusercontent.com/39925111/54237086-74ef9980-4515-11e9-8fe2-00f8de6e7b03.png)
[SOURCE](http://tablaperiodica.uca.es/uca/centro/tablaperiodica/Tabla/elementos/Cobalto/Grupo1/imagenes/radioterapia.jpg)

Throwaway-name commented:
Kevin might want you to mention the nuclear item improvements in the title.

Photoloss commented:
As !FUN! as it would be I don't see how a food irradiator would cook off nukes. [Food irradiation uses ionising radiation](https://en.wikipedia.org/wiki/Food_irradiation) to [smash chemical bonds](https://en.wikipedia.org/wiki/Food_irradiation#Immediate_effects) which generally means bad times for anything trying to be alive but has no effect on atomic nuclei. Nukes and reactors meanwhile operate on [nuclear fission](https://en.wikipedia.org/wiki/Nuclear_fission#Nuclear_Reaction) which generally requires slow neutrons to sustain the chain reaction. Inverse beta decay and spallation via high-energy gamma rays are possible in theory but at the energies available for commercial machines there's absolutely no chance of macroscopic consequences. 

Similar concerns apply to the plutonium cells, and other radioactive material you do not currently mess with. They might get damaged if exposed directly to the primary source for extended periods of time but the same goes for any semiconductor electronics and the sci-fi nanotechnology.

Night-Pryanik commented on src/computer.cpp:
```diff
@@ -1260,6 +1261,169 @@ SHORTLY. TO ENSURE YOUR SAFETY PLEASE FOLLOW THE BELOW STEPS. \n\
+                                g->m.translate_radius( t_water_pool, t_sewage, 8.0, dest, true );
+                                for( const tripoint &area : g->m.points_in_radius( dest, 1 ) ) {
+                                    g->m.add_field( area, fd_nuke_gas, 3 );
+                                }
```
You can replace this snippet here and `g->m.add_field( dest, fd_nuke_gas, 3 );` above with `propagate_field` function.

nexusmrsep replied:
Didn't know it existed, good idea.

Night-Pryanik commented on src/computer.cpp:
```diff
@@ -1260,6 +1261,169 @@ SHORTLY. TO ENSURE YOUR SAFETY PLEASE FOLLOW THE BELOW STEPS. \n\
+                                print_error( _( "EMERGENCY PROCEDURE [1]: EVACUATE! EVACUATE! EAVACUATE!" ) );
+                                add_msg( m_warning, _( "Evacuate Immediately!" ) );
+                                sounds::sound( g->u.pos(), 30, sounds::sound_t::alarm, _( "an alarm sound!" ) );
+                                it->convert( "mininuke_act" );
```
Converting `mininuke_act` into `mininuke_act` is safe?

nexusmrsep replied:
Since (as for below conversation) it will be a conventional explosion, this will be purged anyway. I tested the mininuke, and not tested ```mininuke_act``` assuming it's ok, with changed variables/counters.

Night-Pryanik commented on src/computer.cpp:
```diff
@@ -1260,6 +1261,169 @@ SHORTLY. TO ENSURE YOUR SAFETY PLEASE FOLLOW THE BELOW STEPS. \n\
+            query_any( _( "Conveyor belt cycle complete.  Press any key..." ) );
+            break;
+        }
+        // toggles reinforced glass shutters open->closed and closedd->open depending on their current state
```
```suggestion
        // toggles reinforced glass shutters open->closed and closed->open depending on their current state
```

nexusmrsep commented:
@Photoloss I agree that I might oversimplified radiation decay and it's effects for this purpose, and I'm willing to cut that part off, but Cataclysm wouldn't be Cataclysm without a possibility of some catastrophic events occurring due to players poking their noses where that shouldn't just for the sake of it. Therefore can you hint me on a scenario where such device would go haywire so I could keep the !FUN! part  while being as close to RL constraints as possible? Cobalt-60 is quite radioactive, so some contamination events must be possible if we play it right. 

I was thinking of adding a separate option to extract the radioactive core from the bunker, and option to place some utility bots or another computer option for radiation scrubbers to clean the secondary radiation added by the irradiation process. This itself can be rationalized by the manipulation of radioactive material, as I imagine for in-game purposes that the core is temporarily removed from the water tank to perform it's job and some particles have a chance to spread. Perhaps you can comment on this concept too?

Photoloss commented:
> Therefore can you hint me on a scenario where such device would go haywire so I could keep the !FUN! part while being as close to RL constraints as possible? Cobalt-60 is quite radioactive, so some contamination events must be possible if we play it right.

If you want to explicitly make it run on cobalt the main concern is the primary radiation source. Gamma rays are highly penetrating and cobalt delivers high intensity so if you have less than a thick concrete wall with lead shielding between yourself and the cobalt you're in for a Really Bad Day. A commercial plant would have many safety locks and layers of shielding to prevent exactly that so you would need to hack the facility for overrides and then deliberately walk to the actual radiation source.

Structural damage to the machine could be terrible though. Cobalt "dirty nukes" [have been proposed](https://en.wikipedia.org/wiki/Cobalt-60#Applications) and a non-nuclear explosion near the cobalt storage could easily achieve a similar effect. Again you would have to plant C4 or otherwise deliberately sabotage the system to get a decent yield as a factory would have taken plenty of measures to pre-empt things like gas leaks or electrical faults from causing any issues.

> I was thinking of adding a separate option to extract the radioactive core from the bunker, and option to place some utility bots or another computer option for radiation scrubbers to clean the secondary radiation added by the irradiation process. This itself can be rationalized by the manipulation of radioactive material, as I imagine for in-game purposes that the core is temporarily removed from the water tank to perform it's job and some particles have a chance to spread. Perhaps you can comment on this concept too?

Of course this being Cataclysm's dystopian near-future sci-fi you can implement the entire plant as a robot-operated facility with the manual-access consoles only intended for maintenance. When the entire thing needs to be hacked and reprogrammed for personal use to begin with it's easy to justify potential mistakes in managing the shielding, or you could "accidentally" trigger a cobalt core replacement with no such delivery present (with a half-life of 5 years commercial facilities would definitely need *some* way of replacing their stock). Don't let computers or robots idle near the cobalt itself though as radiation is quite damaging to semiconductor electronics as well, the shields should be opened and robots/staff allowed to enter only when actually necessary.

Since cobalt's activity stems from spontaneous beta decay and water is not particularly good at absorbing gamma rays any process where the source is significantly removed from underground storage during operation would only serve to *reduce* the overall radiation output by bringing it closer instead. It decays into stable nickel so there are no hazardous secondary products either, and unlike neutron-rich fission reactors beta and gamma radiation can't really turn other materials radioactive. 

Unfortunately for you there really isn't much the player could mess up without **intentionally** sabotaging the facility, if you want !FUN! implement a nuclear power plant or something using cryogenic superconducting magnets.

kevingranade commented:
Short of triggering a full detonation, how about having the radiation trigger a misfire instead?
The mechanism would be that the ionizing radiation triggers some of the detonators. Since they would fire out of sequence, the bomb wouldn't go critical, but it would still explode the bomb and spray radioactive materials in all directions.

This could also be applied to other explosives with electrical detonators.

Alternately, if you really like the full detonation option, it can go in CrazyCataclysm no matter how unscientific it is since I think it's fun.

Photoloss commented:
> Short of triggering a full detonation, how about having the radiation trigger a misfire instead?
The mechanism would be that the ionizing radiation triggers some of the detonators. Since they would fire out of sequence, the bomb wouldn't go critical, but it would still explode the bomb and spray radioactive materials in all directions.

Non-nuclear sure. I doubt it would happen *frequently* IRL as detonators should be designed to prevent such failures but sitting in a gamma ray beam isn't the standard test case by any stretch. Accidental triggers of **any** digital technology are possible, in fact semiconductor ionisation is [even used](https://en.wikipedia.org/wiki/Semiconductor_detector#Detection_mechanism) to *detect* gamma rays and particles via principles similar to basic computer components.

I'm not sure what the effect on cartridges and explosives themselves would be. With enough beam intensity you could definitely cause them to combust but I don't know any quick references on how much local energy deposit you actually need.

> Alternately, if you really like the full detonation option, it can go in CrazyCataclysm no matter how unscientific it is since I think it's fun.

In that context I'd think the first course of action would be to transform zombies into Hulks, no?

nexusmrsep commented:
Ok, I'll go for conventional explosion with the containment breach of the radiation source, a.k.a mini dirty bomb. For this I'll merge both catastrophic failures into one. I'll leave behind the idea of ionizing radiation destroying other potential electronic equipment placed on the platform.

For the protocol I'll only mention that overall radiation behavior in game already leans more towards fiction then fact - for example, irradiated player irradiates background? Unless s/he literally carries irradiated mud on the boots, wouldn't it be otherwise impossible?

There is however one more idea I'd like to discuss before implementing. I may code in an option to extract the radiation core, so that the player may interact with it, but I have no idea what would it be useful for. Unless it'd give a bunch of plutonium cells irreversibly shutting down the facility for it's main purpose perhaps?

[SORUCE](https://www.sciencedirect.com/topics/engineering/gamma-radiation)
_The arrays are stored in a deep water pool, which prevents gamma radiation from escaping. The shielding water does not become radioactive.
Large-scale gamma irradiators come in two configurations. In one configuration, the Cobalt arrays are **hoisted out of the water into a radiation chamber** that typically has shields made out of massive concrete or steel. This is shown in Fig. 1.15. The concrete walls are typically 2-feet thick or more. When the irradiators are out of the protective storage pool, all workers are outside the thick walls. Hanging carriers, totes, or roller conveyors are typically employed to move the product through the chamber for irradiation. A typical product container can be 60 cm × 50 cm × 150 cm, and some irradiators are designed to irradiate entire pallets of product measuring 120 cm × 100 cm × 150 cm. Because Cobalt-60 photons have no mass, they can penetrate more than 24 inches (60 cm) of food product if irradiated on both sides.
In the second configuration, unlike a radiation chamber irradiator, an underwater irradiator stores the Cobalt-60 permanently at the bottom of a pool of water. Instead of raising the Cobalt-60 into a shielded chamber, the product, placed in water-free containers, is lowered to the bottom of the pool adjacent to the Cobalt-60 to receive a dose of radiation. No above-ground shielding or radiation chamber is present. Typically, the product is loaded into water-free containers and the containers are lowered/raised using a hoist mechanism._

Nothing better then sweet Cherenkov glow:
![image](https://user-images.githubusercontent.com/39925111/54302492-40391c00-45c1-11e9-8634-8f0d948c7316.png)

This is where I seek the !FUN! part. Many nasty things may happen when a a player messes with exposed core, but with clean suits, hazmat suits, power armor (as radiation protection) and with some iodine tablets this may be utilized as some kind of high risk - high reward game-play opportunity, that I'm reluctant to pass by. Ideas welcomed.

nexusmrsep commented on src/computer.cpp:
```diff
@@ -1260,6 +1261,169 @@ SHORTLY. TO ENSURE YOUR SAFETY PLEASE FOLLOW THE BELOW STEPS. \n\
+            g->u.moves -= 30;
+            bool error = false;
+            for( const tripoint &dest : g->m.points_in_radius( g->u.pos(), 10 ) ) {
+                if( g->m.ter( dest ) == t_floor_blue ) {
```
You are right, I was thinking about it all the time, but now as you said it aloud, I'm convinced.

Rivet-the-Zombie commented on src/computer.cpp:
```diff
@@ -1260,6 +1261,169 @@ SHORTLY. TO ENSURE YOUR SAFETY PLEASE FOLLOW THE BELOW STEPS. \n\
+                                print_error( _( "WARNING [409]: Radiation spike detected!" ) );
+                                print_error( _( "  >> AUTO-RISK-ASSESMENT: Geiger profiling..." ) );
+                                print_error( _( "  >> [Self-sustaining] reaction in [critical mass] detected!\n" ) );
+                                print_error( _( "WARNING [999]: Catastrofic meltdown imminent!\n" ) );
```
Catastrophic

Photoloss commented:
> For this I'll merge both catastrophic failures into one.

Plutonium will not explode from exposure to gamma radiation, only the mininuke has detonators which could misfire. C4 and player-crafted RC bombs can go off at full strength but without radiation.

> For the protocol I'll only mention that overall radiation behavior in game already leans more towards fiction then fact - for example, irradiated player irradiates background? Unless s/he literally carries irradiated mud on the boots, wouldn't it be otherwise impossible?

First of all I'm not sure the player actually *spreads* radiation considering there's a mutation trait dedicated to doing so constantly. Secondly player irradiation is abstracted too much, proximity to reactor slurry or damaged nukes can indeed not only contaminate your clothing but also "activate" material in your body via neutron capture (basically the same mechanism as proposed in the *actual* cobalt dirty bomb)

> There is however one more idea I'd like to discuss before implementing. I may code in an option to extract the radiation core, so that the player may interact with it, but I have no idea what would it be useful for. Unless it'd give a bunch of plutonium cells irreversibly shutting down the facility for it's main purpose perhaps?

Plutonium is **very** different from cobalt! An actual cobalt core can not be used as fission reactor fuel and outputs almost exclusively gamma radiation which has relatively few uses. The level of radiation we are talking about is not sufficient for an "energy beam" weapon and wouldn't even be combat-viable if you strapped a baddie to the core itself.

nexusmrsep commented:
> Plutonium will not explode from exposure to gamma radiation. (...)

That I know. You may misinterpret my intention. I meant to say I will drop the effects on plutonium cells, but use the contamination code from this effect and move it around, so there will be only one coded catastrophic event that would include conventional explosion and contamination. 

> First of all I'm not sure the player actually spreads radiation considering there's a mutation trait dedicated to doing so constantly.

I think you're right. It seems to come from mutation. There again, I could swear something like this has happen in one of Rycon YouTube playthroughs. Memory is a funny thing.

> Plutonium is very different from cobalt!

I know. I'm brainstorming here. I wish to invent some feasible idea for use of the radioactive material in the irradiator core, assuming it will be obtainable by the player.
Option 1:
So if for example it was plutonium cells and not cobalt - the question would be: would an array of plutonium cells (include in-game leeway for the item itself, assuming its existence as axiom) be usable as a radiation source for the food irradiator? If yes, then it would be great, as they are in use, and are in demand. If no - bummer, this way would be closed.
Option 2:
make it cobalt-60, but then - you seem to say that it wouldn't have any use for the player - so in this option you'd propose, all in all, dropping the idea completely?

Photoloss commented:
> So if for example it was plutonium cells and not cobalt - the question would be: would an array of plutonium cells (include in-game leeway for the item itself, assuming its existence as axiom) be usable as a radiation source for the food irradiator? If yes, then it would be great, as they are in use, and are in demand. If no - bummer, this way would be closed.

Using plutonium is a *terrible* idea, precisely for the reasons it is used in nuclear bombs. Plus it is expensive to produce and some isotopes turn into radon which is **not** good.

> Or maybe if cobalt-60 would not have any effect, it might be a potential "death trap" !FUN! - player picks it up "for later use" but all s/he gets is an always radioactive item that should be ditched ASAP?

If you want to put in the work I see no harm in implementing it as a "deathtrap" for now. If nothing else we could use it for on-the-fly food irradiation if we can somehow handle it (technically a hazmat suit would be virtually useless against gamma radiation but in-game it completely protects you from everything). In the (far) future it might be used in improvised medical treatments or for the confirmed sci-fi effects.

nexusmrsep commented:
Should be done. I've applied the results of discussion.

Night-Pryanik commented on src/computer.cpp:
```diff
@@ -1260,6 +1261,172 @@ SHORTLY. TO ENSURE YOUR SAFETY PLEASE FOLLOW THE BELOW STEPS. \n\
+            break;
+        }
+
+        // imitates item movement through conveyor belt throgh 3 different loading/unloading bays
```
```suggestion
        // imitates item movement through conveyor belt through 3 different loading/unloading bays
```

kevingranade commented:
No map location for this at all? I need this spawning somewhere so I can test it.

nexusmrsep commented:
Guess that means I got to learn new skill: mapgen. But that's fair, computer can't be spawned just like that to test it functions properly. I'll make a food processing plant, or lab finale. Or both.

Night-Pryanik commented on data/json/items/chemicals_and_resources.json:
```diff
@@ -1157,6 +1157,20 @@
+    "name": "cobalt-60 pellet",
+    "symbol": "=",
+    "color": "green",
+    "description": "Radioactive material that used to be a part of some nuclear industry equipment. You are yet to find some use for it.",
```
```suggestion
    "description": "Radioactive material that used to be a part of some nuclear industry equipment.  You are yet to find some use for it.",
```

Night-Pryanik commented on data/json/items/generic.json:
```diff
@@ -2842,6 +2842,21 @@
+    "symbol": ",",
+    "color": "green",
+    "name": "industrial ID card",
+    "description": "This ID card once belonged to a high level techincian.  The reverse side describes protocol for using it; this could grant access at one control panel, if you can find one.",
```
```suggestion
    "description": "This ID card once belonged to a high level technician.  The reverse side describes protocol for using it; this could grant access at one control panel, if you can find one.",
```

nexusmrsep commented:
@kevingranade 
As requested I've built a fully functional facility overmap location with many fun features.
Skill: mapgen creation achieved.

Night-Pryanik commented on data/json/mapgen/irradiator_1.json:
```diff
@@ -0,0 +1,479 @@
+      { "item": "rag_bloody", "prob": 30 },
+      { "group": "drugs_emergency", "prob": 30 },
+      { "group": "drugs_heal_simple", "prob": 20 },
+      { "gropu": "drugs_analgesis", "prob": 10 }
```
```suggestion
      { "gropu": "drugs_analgesic", "prob": 10 }
```

Night-Pryanik commented on data/json/mapgen/irradiator_1.json:
```diff
@@ -0,0 +1,479 @@
+            { "name": "Cycle conveyor belt", "action": "conveyor", "security": 0 },
+            { "name": "Toggle safety shutters", "action": "shutters", "security": 0 },
+            { "name": "Probe radiation levels", "action": "geiger", "security": 0 },
+            { "name": "Comence irradiation sequence", "action": "irradiator", "security": 0 },
```
```suggestion
            { "name": "Commence irradiation sequence", "action": "irradiator", "security": 0 },
```

Night-Pryanik commented on data/json/mapgen/irradiator_1.json:
```diff
@@ -0,0 +1,479 @@
+          "name": "Power Management",
+          "security" : 2,
+          "options": [
+            { "name": "External power management.", "action": "disconnect", "security": 0 },
```
```suggestion
            { "name": "External power management", "action": "disconnect", "security": 0 },
```

Night-Pryanik commented on data/json/mapgen/irradiator_1.json:
```diff
@@ -0,0 +1,479 @@
+          "security" : 2,
+          "options": [
+            { "name": "External power management.", "action": "disconnect", "security": 0 },
+            { "name": "Backup power management.", "action": "disconnect", "security": 0 },
```
```suggestion
            { "name": "Backup power management", "action": "disconnect", "security": 0 },
```

Night-Pryanik commented on data/json/mapgen/irradiator_1.json:
```diff
@@ -0,0 +1,479 @@
+          "options": [
+            { "name": "External power management.", "action": "disconnect", "security": 0 },
+            { "name": "Backup power management.", "action": "disconnect", "security": 0 },
+            { "name": "Run emergency auto-diagnostic.", "action": "deactivate_shock_vent", "security": 2 }
```
```suggestion
            { "name": "Run emergency auto-diagnostic", "action": "deactivate_shock_vent", "security": 2 }
```

Night-Pryanik commented on src/computer.cpp:
```diff
@@ -1262,6 +1263,233 @@ SHORTLY. TO ENSURE YOUR SAFETY PLEASE FOLLOW THE STEPS BELOW. \n\
+        }
+        // toggles reinforced glass shutters open->closed and closed->open depending on their current state
+        case COMPACT_SHUTTERS:
+            g->u.moves -= 300;
```
The same here.

Night-Pryanik commented on src/computer.cpp:
```diff
@@ -1262,6 +1263,233 @@ SHORTLY. TO ENSURE YOUR SAFETY PLEASE FOLLOW THE STEPS BELOW. \n\
+        // extract radiation source material from irradiator
+        case COMPACT_EXTRACT_RAD_SOURCE:
+            if( query_yn( _( "Operation irreversible.  Extract radioactive material?" ) ) ) {
+                g->u.moves -= 300;
```
The same here.

Night-Pryanik commented on src/computer.cpp:
```diff
@@ -1262,6 +1263,233 @@ SHORTLY. TO ENSURE YOUR SAFETY PLEASE FOLLOW THE STEPS BELOW. \n\
+                    g->m.translate_radius( t_rad_platform, t_concrete, 8.0, g->u.pos(), true );
+                    remove_option( COMPACT_IRRADIATOR );
+                    remove_option( COMPACT_EXTRACT_RAD_SOURCE );
+                    query_any( _( "Extraction sequence complete... press any key." ) );
```
```suggestion
                    query_any( _( "Extraction sequence complete... Press any key." ) );
```

Night-Pryanik commented on src/computer.cpp:
```diff
@@ -1262,6 +1263,233 @@ SHORTLY. TO ENSURE YOUR SAFETY PLEASE FOLLOW THE STEPS BELOW. \n\
+                    remove_option( COMPACT_EXTRACT_RAD_SOURCE );
+                    query_any( _( "Extraction sequence complete... press any key." ) );
+                } else {
+                    query_any( _( "ERROR!  Radiation platform unresponsive... press any key." ) );
```
```suggestion
                    query_any( _( "ERROR!  Radiation platform unresponsive... Press any key." ) );
```

Night-Pryanik commented on src/computer.cpp:
```diff
@@ -1262,6 +1263,233 @@ SHORTLY. TO ENSURE YOUR SAFETY PLEASE FOLLOW THE STEPS BELOW. \n\
+            }
+            break;
+        // remove shock vent fields; check for existing plutonium generators in radius
+        case COMPACT_DEACTIVATE_SHOCK_VENT:
```
There's no move cost at all?

nexusmrsep replied:
Right, fixing.

Night-Pryanik commented on src/computer.cpp:
```diff
@@ -1262,6 +1263,233 @@ SHORTLY. TO ENSURE YOUR SAFETY PLEASE FOLLOW THE STEPS BELOW. \n\
+            }
+            print_line( _( "Initiating POWER-DIAG ver.2.34 ..." ) );
+            if( has_vent ) {
+                print_error( _( "Short circuit detected!." ) );
```
```suggestion
                print_error( _( "Short circuit detected!" ) );
```

Night-Pryanik commented on src/iuse.cpp:
```diff
@@ -5301,8 +5302,8 @@ int iuse::radglove( player *p, item *it, bool, const tripoint & )
             p->add_msg_player_or_say( m_info,
-                                      _( "Your radiation level: %d" ),
-                                      _( "It says here that my radiation level is %d" ),
+                                      _( "Your radiation level: %d mSv" ),
```
```suggestion
                                      _( "Your radiation level: %d mSv." ),
```

Night-Pryanik commented on src/iuse.cpp:
```diff
@@ -5301,8 +5302,8 @@ int iuse::radglove( player *p, item *it, bool, const tripoint & )
-                                      _( "Your radiation level: %d" ),
-                                      _( "It says here that my radiation level is %d" ),
+                                      _( "Your radiation level: %d mSv" ),
+                                      _( "It says here that my radiation level is %d mSv" ),
```
```suggestion
                                      _( "It says here that my radiation level is %d mSv." ),
```

Night-Pryanik commented on src/iexamine.cpp:
```diff
@@ -704,7 +704,7 @@ void iexamine::cardreader( player &p, const tripoint &examp )
     bool open = false;
     itype_id card_type = ( g->m.ter( examp ) == t_card_science ? "id_science" :
-                           "id_military" );
+                           g->m.ter( examp ) == t_card_military ? "id_military" : "id_industrial" );
```
Won't this change break science card reader functionality?

nexusmrsep replied:
I hope not. This was meant to expand the switch between different card types. Will double check ASAP.

nexusmrsep replied:
A quick test showed that it doesn't break neither the science nor military card ID readers.

nexusmrsep commented on src/computer.cpp:
```diff
@@ -1262,6 +1263,233 @@ SHORTLY. TO ENSURE YOUR SAFETY PLEASE FOLLOW THE STEPS BELOW. \n\
+        // imitates item movement through conveyor belt through 3 different loading/unloading bays
+        // ensure only bay of each type in range
+        case COMPACT_CONVEYOR: {
+            g->u.moves -= 300;
```
Pressing the button takes no time, but conveyor belt needs time to physically move items few tiles. I can't magically compress time here. Same goes for the following move consumption computer actions. Is there a better way to represent time passage needed for the mechanical aspects of.those actions?

Night-Pryanik replied:
I believe there isn't, unless you add a prolonged activity mechanics for this new furniture - the one like washing machine or smoker have.

nexusmrsep replied:
I don't want to do any out-of-console things that would look awkward to the player to not confuse anyone. The best option would be to either:
1] make some fake "progress bar" in console that would show that "something is happening, we are waiting for it to finish"
2] introducing functionality that would actually move items one by one from point A to point B using a conveyor belt tiles as a pathfinding aid, that would be an independent from player's actions. 
Both are too far from either scope of this PR and my skill level.

kevingranade commented:
Needs styling, but checking it out otherwise.

kevingranade commented:
Crashed when I cycled the conveyor belt
```
MESSAGE: SIGSEGV: Segmentation fault
STACK TRACE:

        ./cataclysm(_Z21debug_write_backtraceRSo+0x25) [0x7ac6eb]
        ./cataclysm() [0x79fb7c]
        ./cataclysm() [0x79fd7d]
        /lib64/libc.so.6(+0x347e0) [0x7f63771897e0]
        ./cataclysm(_ZN4itemC2ERKS_+0x18) [0x6cf054]
        ./cataclysm(_ZNSt7__cxx114listI4itemSaIS1_EE14_M_create_nodeIJRKS1_EEEPSt10_List_nodeIS1_EDpOT_+0x26) [0x6cf4de]
        ./cataclysm(_ZNSt7__cxx114listI4itemSaIS1_EEC1ERKS3_+0x2d) [0x6cf529]
        ./cataclysm(_ZN4itemC2ERKS_+0x30) [0x6cf06c]
        ./cataclysm(_ZN8computer17activate_functionE15computer_action+0x3fbe) [0x76c4dc]
        ./cataclysm(_ZN8computer3useEv+0x494) [0x76d36c]
        ./cataclysm(_ZN4game12use_computerERK8tripoint+0x273) [0x837395]
        ./cataclysm(_ZN4game7examineERK8tripoint+0x14a) [0x84ddd4]
        ./cataclysm(_ZN4game7examineEv+0xe5) [0x84e233]
        ./cataclysm(_ZN4game13handle_actionEv+0x169e) [0x87fbda]
        ./cataclysm(_ZN4game7do_turnEv+0x3e6) [0x849d54]
        ./cataclysm(main+0x1152) [0x68792f]
        /lib64/libc.so.6(__libc_start_main+0xea) [0x7f637717602a]
        ./cataclysm(_start+0x2a) [0x6b224a]

        Attempting to repeat stack trace using debug symbols...
        debug_write_backtrace(std::ostream&)
        .../src/debug.cpp:593
        std::__cxx11::basic_ostringstream<char, std::char_traits<char>, std::allocator<char> >::str() const
        /usr/include/c++/7/sstream:638
        log_crash
        .../src/crash.cpp:258
        signal_handler
        .../src/crash.cpp:287
        ??
        ??:0
        item::item(item const&)
        .../src/item.h:175
        std::_List_node<item>* std::__cxx11::list<item, std::allocator<item> >::_M_create_node<item const&>(item const&)
        /usr/include/c++/7/bits/stl_list.h:579
        void std::__cxx11::list<item, std::allocator<item> >::_M_insert<item const&>(std::_List_iterator<item>, item const&)
        /usr/include/c++/7/bits/stl_list.h:1802
        void std::__cxx11::list<item, std::allocator<item> >::emplace_back<item const&>(item const&)
        /usr/include/c++/7/bits/stl_list.h:1133
        void std::__cxx11::list<item, std::allocator<item> >::_M_initialize_dispatch<std::_List_const_iterator<item> >(std::_List_const_iterator<item>, std::_List_const_iterator<item>, std::__false_type)
        /usr/include/c++/7/bits/stl_list.h:1730
        std::__cxx11::list<item, std::allocator<item> >::list(std::__cxx11::list<item, std::allocator<item> > const&)
        /usr/include/c++/7/bits/stl_list.h:655
        item::item(item const&)
        .../src/item.h:175
        computer::activate_function(computer_action)
        .../src/computer.cpp:1424
        computer::use()
        .../src/computer.cpp:227
        game::use_computer(tripoint const&)
        .../src/game.cpp:5069
        game::examine(tripoint const&)
        .../src/game.cpp:6151
        game::examine()
        .../src/game.cpp:6062
        game::handle_action()
        .../src/handle_action.cpp:1554
        game::do_turn()
        .../src/game.cpp:1401
        main
        .../src/main.cpp:682
        __libc_start_main
        ??:?
        _start
        ??:?
```

nexusmrsep commented:
Never happened to me. Will look into it. Any specific cause on the glance
of an eye that could have caused it?

pt., 19 kwi 2019, 07:53 użytkownik Kevin Granade <notifications@github.com>
napisał:

> Crashed when I cycled the conveyor belt
>
> MESSAGE: SIGSEGV: Segmentation fault
> STACK TRACE:
>
>         ./cataclysm(_Z21debug_write_backtraceRSo+0x25) [0x7ac6eb]
>         ./cataclysm() [0x79fb7c]
>         ./cataclysm() [0x79fd7d]
>         /lib64/libc.so.6(+0x347e0) [0x7f63771897e0]
>         ./cataclysm(_ZN4itemC2ERKS_+0x18) [0x6cf054]
>         ./cataclysm(_ZNSt7__cxx114listI4itemSaIS1_EE14_M_create_nodeIJRKS1_EEEPSt10_List_nodeIS1_EDpOT_+0x26) [0x6cf4de]
>         ./cataclysm(_ZNSt7__cxx114listI4itemSaIS1_EEC1ERKS3_+0x2d) [0x6cf529]
>         ./cataclysm(_ZN4itemC2ERKS_+0x30) [0x6cf06c]
>         ./cataclysm(_ZN8computer17activate_functionE15computer_action+0x3fbe) [0x76c4dc]
>         ./cataclysm(_ZN8computer3useEv+0x494) [0x76d36c]
>         ./cataclysm(_ZN4game12use_computerERK8tripoint+0x273) [0x837395]
>         ./cataclysm(_ZN4game7examineERK8tripoint+0x14a) [0x84ddd4]
>         ./cataclysm(_ZN4game7examineEv+0xe5) [0x84e233]
>         ./cataclysm(_ZN4game13handle_actionEv+0x169e) [0x87fbda]
>         ./cataclysm(_ZN4game7do_turnEv+0x3e6) [0x849d54]
>         ./cataclysm(main+0x1152) [0x68792f]
>         /lib64/libc.so.6(__libc_start_main+0xea) [0x7f637717602a]
>         ./cataclysm(_start+0x2a) [0x6b224a]
>
>         Attempting to repeat stack trace using debug symbols...
>         debug_write_backtrace(std::ostream&)
>         .../src/debug.cpp:593
>         std::__cxx11::basic_ostringstream<char, std::char_traits<char>, std::allocator<char> >::str() const
>         /usr/include/c++/7/sstream:638
>         log_crash
>         .../src/crash.cpp:258
>         signal_handler
>         .../src/crash.cpp:287
>         ??
>         ??:0
>         item::item(item const&)
>         .../src/item.h:175
>         std::_List_node<item>* std::__cxx11::list<item, std::allocator<item> >::_M_create_node<item const&>(item const&)
>         /usr/include/c++/7/bits/stl_list.h:579
>         void std::__cxx11::list<item, std::allocator<item> >::_M_insert<item const&>(std::_List_iterator<item>, item const&)
>         /usr/include/c++/7/bits/stl_list.h:1802
>         void std::__cxx11::list<item, std::allocator<item> >::emplace_back<item const&>(item const&)
>         /usr/include/c++/7/bits/stl_list.h:1133
>         void std::__cxx11::list<item, std::allocator<item> >::_M_initialize_dispatch<std::_List_const_iterator<item> >(std::_List_const_iterator<item>, std::_List_const_iterator<item>, std::__false_type)
>         /usr/include/c++/7/bits/stl_list.h:1730
>         std::__cxx11::list<item, std::allocator<item> >::list(std::__cxx11::list<item, std::allocator<item> > const&)
>         /usr/include/c++/7/bits/stl_list.h:655
>         item::item(item const&)
>         .../src/item.h:175
>         computer::activate_function(computer_action)
>         .../src/computer.cpp:1424
>         computer::use()
>         .../src/computer.cpp:227
>         game::use_computer(tripoint const&)
>         .../src/game.cpp:5069
>         game::examine(tripoint const&)
>         .../src/game.cpp:6151
>         game::examine()
>         .../src/game.cpp:6062
>         game::handle_action()
>         .../src/handle_action.cpp:1554
>         game::do_turn()
>         .../src/game.cpp:1401
>         main
>         .../src/main.cpp:682
>         __libc_start_main
>         ??:?
>         _start
>         ??:?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/CleverRaven/Cataclysm-DDA/pull/28669#issuecomment-484775164>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AJQTK547OUOVOFUUHKWJNZDPRFM5BANCNFSM4G5PWCKA>
> .
>

nexusmrsep commented:
I seems that on some point I've changed the item removal in conveyor belt processing not noticing it's placed inside a loop and without retesting. 
Fixed now.
Also noticed that potatoes weren't being irradiated properly because I've migrated only half of the naming scheme.
Also fixed.

ZhilkinSerg commented on src/computer.cpp:
```diff
@@ -1262,6 +1263,234 @@ SHORTLY. TO ENSURE YOUR SAFETY PLEASE FOLLOW THE STEPS BELOW. \n\
+                    sum_rads += g->m.get_radiation( platform );
+                    tiles_counted ++;
+                    if( g->m.get_radiation( platform ) > peak_rad ) {
+                        peak_rad = g->m.get_radiation( dest );
```
Shouldn't `platform` be used here instead of `dest`?

nexusmrsep replied:
True. Fixing.
